src/coprocessor/dag/handler.rs

```diff
@@ -32,12 +32,61 @@ pub struct DAGRequestHandler {
 }
 
 impl DAGRequestHandler {
-    pub fn build<S: Store + 'static>(
+    fn build_dag<S: Store + 'static>(
+        eval_cfg: EvalConfig,
         mut req: DAGRequest,
         ranges: Vec<KeyRange>,
         store: S,
         deadline: Deadline,
         batch_row_limit: usize,
+    ) -> Result<Self> {
+        let executor = super::builder::DAGBuilder::build_normal(
+            req.take_executors().into_vec(),
+            store,
+            ranges,
+            Arc::new(eval_cfg),
+            req.get_collect_range_counts(),
+        )?;
+        Ok(Self {
+            deadline,
+            executor,
+            output_offsets: req.take_output_offsets(),
+            batch_row_limit,
+        })
+    }
+
+    fn build_batch_dag<S: Store + 'static>(
```

Comment:
I don't think it's possible, because batch executor and non-batch executors are different executors, and they use different DAG handlers. Once executors are built, there is no way to switch to another (unless you rebuild it). I suggest to decide this before building. There is a flag called `enable_batch_if_possible` that can forcely disables batch execution dynamically in future. See @AndreMouche 's comment.